sale_order_import_edifact/wizard/sale_order_import.py

```diff
@@ -0,0 +1,277 @@
+    file_ext = fields.Char(default="txt,d96a")
+    # EDIFACT format release
+    release = fields.Char(default="d96a")
+
+    @api.depends("order_file", "order_filename")
+    def _compute_edifact_ok(self):
+        for rec in self:
+            if not (rec.order_file and rec.order_filename):
+                rec.edifact_ok = False
+                continue
+            extensions = rec.file_ext.split(",")
+            path, ext = os.path.splitext(self.order_filename)
+            ok = ext and ext[1:] in extensions
+            if not ok:
+                ok = b64decode(self.order_file[:4]) == b"UNB"
+            rec.edifact_ok = ok
+
+    def button_parse_order_preview(self):
+        """
+        Generate a preview of the data before importing.
+
+        This method is called by the button to take a quick look at the pydifact
+        structure ('edifact' type) and the parsed object ('order') from
+        the original document before importing the document.
+        """
+
+        self.ensure_one()
+        edifact_model = self.env["base.edifact"]
+        order_file_decoded = b64decode(self.order_file)
+        parsed_obj = dict(edifact=None, order=None)
+        if self.edifact_ok:
+            parsed_obj["edifact"] = edifact_model.pydifact_obj(order_file_decoded)
+
+        parsed_order = self.parse_order(
+            order_file_decoded, self.order_filename, self.partner_id
+        )
+        del parsed_order["attachments"]
+        parsed_obj["order"] = parsed_order
+        self.write(
+            {
+                "json_data_preview": json.dumps(parsed_obj, indent=4, default=str),
+            }
+        )
+        action = self.env["ir.actions.act_window"]._for_xml_id(
+            "sale_order_import.sale_order_import_action"
+        )
+        action["res_id"] = self.id
+        return action
+
+    def _get_supported_types(self):
+        # Add more types for EDIFACT
+        res = super()._get_supported_types()
+        res.update({"edifact": ("text/plain")})
+        return res
+
+    @api.model
+    def parse_edifact_order(self, filecontent, detect_doc_type=False):
+        # https://github.com/nerdocs/pydifact/blob/master/pydifact/segmentcollection.py
+        if detect_doc_type:
+            if not self.edifact_ok:
+                return None
+            return self.env.context.get("doc_type", "rfq")
+
+        edifact_model = self.env["base.edifact"]
+        interchange = edifact_model._loads_edifact(filecontent)
+        header = interchange.get_header_segment()
+        # > UNB segment: [['UNOA', '2'], ['5450534000000', '14'],
+        # ['8435337000003', '14'], ['230306', '0435'], '5506']
+
+        msg_type, msg_type_release = edifact_model._get_msg_type(interchange)
+        supported = ["ORDERS"]
+        if msg_type not in supported:
+            raise UserError(_("%s document is not a Sale Order document"))
+
+        doc_type = self.env.context.get("doc_type", "order")
```

Comment:
default_doc_type?
Although, at this stage `self.doc_type` should be set properly no?